api/v1beta1/cinder_types.go

```diff
@@ -17,57 +17,130 @@ limitations under the License.
+	// +kubebuilder:default=false
+	// PreserveJobs - do not delete jobs after they finished e.g. to check logs
+	PreserveJobs bool `json:"preserveJobs,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="# add your customization here"
+	// CustomServiceConfig - customize the service config for all Cinder services using this parameter to change service defaults,
+	// or overwrite rendered information using raw OpenStack config format. The content gets added to
+	// to /etc/<service>/<service>.conf.d directory as custom.conf file.
+	CustomServiceConfig string `json:"customServiceConfig,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// ConfigOverwrite - interface to overwrite default config files like e.g. logging.conf or policy.json.
+	// But can also be used to add additional files. Those get added to the service config dir in /etc/<service> .
+	// TODO: -> implement
+	DefaultConfigOverwrite map[string]string `json:"defaultConfigOverwrite,omitempty"`
+
+	// TODO: We will need to decide which fields within the specs below we want to
+	// be required or optional.  We probably also want to use webhooks and/or kubebuilder
+	// defaults to set the fields when they are not provided by the user
+	// TODO: As we flesh out functionality in the operator, we will need
+	// to address these fields' optional/required/default considerations given
+	// that CinderAPI, CinderBackup, CinderScheduler and CinderVolume are
+	// intended to have their Specs embedded within this parent Cinder CRD.
+	// Becuse we are embedding their specs, all admission field verification will
+	// fire (as if an actual CR of one of the types had been applied/created).
+	// We have to find the right balance between what we would expect users
+	// to provide in the Cinder CR for the childrens' Specs versus what we
+	// provide via inheriting from the Cinder CR or through webhooks, defaults, etc
+
+	// +kubebuilder:validation:Required
+	// CinderAPI - Spec definition for the API service of this Cinder deployment
+	CinderAPI CinderAPISpec `json:"cinderAPI"`
 
-// Volume defines cinder volume configuration parameters
-type Volume struct {
-	// Name of cinder volume service
-	Name string `json:"name,omitempty"`
-	// Cinder Volume Container Image URL
-	CinderVolumeContainerImage string `json:"cinderVolumeContainerImage,omitempty"`
-	// Cinder Volume Replicas
-	CinderVolumeReplicas int32 `json:"cinderVolumeReplicas"`
-	// Cinder Volume node selector
-	CinderVolumeNodeSelectorRoleName string `json:"cinderVolumeNodeSelectorRoleName,omitempty"`
+	// +kubebuilder:validation:Required
+	// CinderScheduler - Spec definition for the Scheduler service of this Cinder deployment
+	CinderScheduler CinderSchedulerSpec `json:"cinderScheduler"`
+
+	// +kubebuilder:validation:Required
+	// CinderBackup - Spec definition for the Backup service of this Cinder deployment
+	CinderBackup CinderBackupSpec `json:"cinderBackup"`
+
+	// +kubebuilder:validation:Required
```

Comment:
you could have no volume service ... might not make sense, but it would be ok. so not sure if this should be required or not.

Reply:
Fair enough.  Let's make it optional for now and the storage experts can tweak it if necessary later on.